modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+  extern const CHPL_RT_MD_STR_COPY_DATA: chpl_mem_descInt_t;
+
+  inline proc chpl_string_comm_get(dest: bufferType, src_loc_id: int(64),
+                                   src_addr: bufferType, len: size_t) {
+    __primitive("chpl_comm_get", dest, src_loc_id, src_addr, len);
   }
 
-  inline proc +(s: c_string, x: string) {
-    var cs = __primitive("string_concat", s, x.c_str());
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  proc copyRemoteBuffer(src_loc_id: int(64), src_addr: bufferType, len: int): bufferType {
+      const dest = chpl_mem_alloc((len+1).safeCast(size_t), CHPL_RT_MD_STR_COPY_REMOTE): bufferType;
+      chpl_string_comm_get(dest, src_loc_id, src_addr, len.safeCast(size_t));
+      dest[len] = 0;
+      return dest;
   }
 
-  inline proc +(s: string, x: c_string) {
-    var cs = __primitive("string_concat", s.c_str(), x);
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  extern proc memmove(destination: c_ptr, const source: c_ptr, num: size_t);
+
+  // pointer arithmetic used for strings
+  inline proc +(a: c_ptr, b: integral) return __primitive("+", a, b);
+
+  config param debugStrings = false;
+}
+
+// Chapel Strings
+module String {
+  use BaseStringType;
+  use StringCasts;
+
+  inline proc chpl_debug_string_print(s: c_string) {
+    if debugStrings {
+      extern proc printf(format: c_string, x...);
+      printf("%s\n", s);
+    }
   }
 
-  inline proc +(s: string, x: numeric)
-    return s + x:string;
+  // TODO: We should be able to remove "ignore noinit", but doing so causes
+  // memory leaks in various places. Investigate why later and add noinit back
+  // in when possible.
+  pragma "string record"
+  pragma "ignore noinit"
+  pragma "no default functions"
+  record string {
+    var len: int = 0; // length of string in bytes
+    var _size: int = 0; // size of the buffer we own
+    var buff: bufferType = nil;
+    var owned: bool = true;
+
+    proc string(s: string, owned: bool = true) {
+      this.owned = owned;
+      const sRemote = s.locale.id != chpl_nodeID;
+      if !owned && sRemote {
+        // TODO: Should I just ignore the supplied value of owned instead of
+        //       halting?
+        halt("Must take ownership of a copy of a remote string");
+      }
+      var localBuff: bufferType = nil;
+      const sLen = s.len;
+      // Dont need to do anything if s is an empty string
+      if sLen != 0 {
+        localBuff = if sRemote
+          then copyRemoteBuffer(s.locale.id, s.buff, sLen)
+          else s.buff;
+      }
+      // We only need to copy when the buffer is local
+      // TODO: reinitString does more checks then we actually need but is still
+      // correct. Is it worth just doing the proper copy here?
+      this.reinitString(localBuff, sLen, sLen+1, needToCopy=sRemote);
```

Comment:
I'm really confused by needToCopy=sRemote.
Shouldn't it be needToCopy = ! sRemote ?
I think this code would be cleanest if you created also copyLocalBuffer and then filled in localBuff from that
if a copy was needed.

I don't understand why a copy would be needed (except for the remote case). Is it if owned=true then we need to copy the buffer from the other string, right? (but I'm not sure I'm on board with string being mutable..) Do we need to copy if the s argument was owned, because it might be freed at a different time? I think that calling reinitString makes these issues *less* apparent, so I think the code would be better if you did the right thing directly, here.